text/0000-thread-safe-handles.md

```diff
@@ -0,0 +1,362 @@
+
+    /// Decrements the strong count on the underlying threadsafe function.
+    /// When the count reaches zero, the event loop will not be kept running.
+    /// An unreferenced `EventQueue` may not execute submitted functions.
+    /// _Idempotent_
+    pub fn unref(&mut self);
+
+    /// Increments the strong count on the underlying function.
+    /// If the count is greater than zero, the event loop will be kept running.
+    /// _Idempotent_
+    pub fn ref(&mut self);
+
+    /// Schedules a closure to execute on the main JavaScript thread
+    /// Blocks if the event queue is full
+    pub fn send(
+        &self,
+        f: impl FnOnce(TaskContext) -> NeonResult<()>,
+    ) -> Result<(), EventQueueError>;
+
+    /// Schedules a closure to execute on the main JavaScript thread
+    /// Non-blocking
+    pub fn try_send(
+        &self,
+        f: impl FnOnce(TaskContext) -> NeonResult<()>,
+    ) -> Result<(), EventQueueError>;
+}
+```
+
+The native `napi_call_threadsafe_function` can fail in three ways:
+
+* `napi_queue_full` if the queue is full
+* `napi_invalid_arg` if the thread count is zero.ß This is due to misuse and statically enforced by ownership rules and the `Drop` trait.
+* `napi_generic_failure` if the call to `uv_async_send` fails
+
+These three failures may be reduced to two and are represented by the `EventQueueError<F>` enum.
+
+```rust
+enum EventQueueError<F> {
+    /// The bounded queue was full. The closure is returned in the error.
+    QueueFull(F),
+    /// A generic failure. Most likely on `uv_async_send(&async)`.
+    Unknown,
+}
+```
+
+In the event of an `napi_invalid_arg` error, neon will `panic` as this indicates a bug in neon and not an expected error condition.
+
+#### Design Notes
+
+The backing `napi_threadsafe_function` is already `Send` and `Sync`, allowing the use of `send` with a shared reference.
+
+It is possible to implement `Clone` for `EventQueue` using `napi_acquire_threadsafe_function`; however, it was considered more idiomatic and flexible to leave this to users with Rust reference counting. E.g., `Arc<EventQueue>`.
+
+Both `ref` and `unref` mutate the behavior of the underlying threadsafe function and require exclusive references. These methods are infallible because Rust is capable of statically enforcing the invariants. We may want to optimize with `assert_debug!` on the `napi_status`.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+![standards](https://imgs.xkcd.com/comics/standards.png)
+
+Neon already has `Task`, `EventHandler`, a [proposed](https://github.com/neon-bindings/rfcs/pull/30) `TaskBuilder` and an accepted, but currently unimplemented, [update](https://github.com/neon-bindings/rfcs/pull/28) to `EventHandler`. This is a large amount of API surface area without clear indication of what a user should use.
+
+This can be mitigated with documentation and soft deprecation of existing methods as we get a clearer picture of what a high-level, ergonomic API would look like.
```

Comment:
Worth also clarifying that `Task` and `TaskBuilder` can be collapsed into one thing, and represents access to the Node thread pool, which is orthogonal to sharing handles across threads.